src/lib/logical_query_plan/abstract_lqp_node.cpp

```diff
@@ -445,33 +438,32 @@ std::shared_ptr<AbstractLQPNode> AbstractLQPNode::_shallow_copy(LQPNodeMapping&
 }
 
 void AbstractLQPNode::_remove_output_pointer(const AbstractLQPNode& output) {
+  // We need to lock here because two outputs can be deallocated at the same time, leading to a data race.
+  const auto lock = std::lock_guard{_output_mutex};
   const auto iter = std::ranges::find_if(_outputs, [&](const auto& other) {
     /**
      * HACK!
-     *  Normally we'd just check `&output == other.lock().get()` here.
-     *  BUT (this is the hacky part), we're checking for `other.expired()` here as well and accept an expired element as
-     *  a match. If nothing else breaks the only way we might get an expired element is if `other` is the
-     *  expired weak_ptr<> to `output` - and thus the element we're looking for - in the following scenario:
+     *  Normally, we would just check `&output == other.lock().get()` here.
+     *  BUT (this is the hacky part), we are checking for `other.expired()` here as well and accept an expired element
+     *  as a match. If nothing else breaks, the only way we might get an expired element is if `other` is the
+     *  expired weak_ptr<> to `output` - and, thus, the element we are looking for - in the following scenario:
      *
-     * auto node_a = Node::make()
-     * auto node_b = Node::make(..., node_a)
+     *    auto node_a = Node::make();
+     *    auto node_b = Node::make(..., node_a);
      *
-     * node_b.reset(); // node_b::~AbstractLQPNode() will call `node_a.remove_output_pointer(node_b)`
-     *                 // But we cannot lock node_b anymore, since its ref count is already 0.
+     *    node_b.reset(); // `node_b::~AbstractLQPNode()` will call `node_a._remove_output_pointer(*node_b)`. However,
+     *                    // we cannot lock node_b anymore: its reference count is already 0.
```

Comment:
Given the three stated lines here: why would the ref counter be 0? It's still on the test stack.